lib/pbench/server/filetree.py

```diff
@@ -903,3 +903,24 @@ def delete(self, dataset: str):
         tarball.controller.delete(dataset)
         del self.datasets[dataset]
         self._clean_empties(tarball.controller_name)
+
+    def find_inventory(self, dataset: str) -> Path:
+        """
+        Given the name of a dataset,search the INCOMING tree directory for
+        the dataset and return its relative path.
+
+        Args:
+            dataset: The name of a dataset that might exist somewhere in the
+                file tree
+
+        Raises:
+            DatasetNotFound: the INCOMING tree does not contain a dataset ,either
+            it is not unpacked or dataset is not found .
+
+        Returns:
+            A Path of the dataset directory.
+        """
+        for dirs in self.incoming_root.rglob("*"):
```

Comment:
This will find *controllers* rather than *datasets*; e.g., `dhcp31-194.perf.lab.eng.bos.redhat.com` rather than `pbench-user-benchmark_loop.3_2022.06.14T20.38.11`

Also, while right now `dataset.name` will match the tarball name, and I'll need to change some aspects of `FileTree` anyway when I change that, I'd rather not build in another instance that needs to be fixed later.

My idea had been to use `file_tree.find_dataset(dataset.name)` instead. If the dataset exists in the file tree, you'll get back a `Tarball`, and `tarball.unpacked` is the `Path` to the unpacked inventory if it exists. Then you can just search from that `Path` for the file you want.

Reply:
I was confused here; since your API actually works on your test server, I had to come back and look again. And, yeah, you're using `rglob`, which is recursive, so it will iterate through *the full file tree of every unpacked tarball in every controller* under the `incoming` base directory until it finds a `.name` matching the dataset. (And although an inappropriate match is probably unlikely, this code doesn't even care whether the match is a directory, and it probably should.) Regardless, this could take a long time and waste a lot of resources.

There are better ways to code this directly; however, again, using `.find_dataset` to locate the named dataset and looking at the `.unpacked` attribute on the object instance would be a more consistent solution.